dangerzone/conversion/common.py

```diff
@@ -0,0 +1,140 @@
+TIMEOUT_MIN: float = 60  # (seconds)
+
+
+def running_on_qubes() -> bool:
+    # https://www.qubes-os.org/faq/#what-is-the-canonical-way-to-detect-qubes-vm
+    return os.path.exists("/usr/share/qubes/marker-vm")
+
+
+async def read_stream(
+    sr: asyncio.StreamReader, callback: Optional[Callable] = None
+) -> bytes:
+    """Consume a byte stream line-by-line.
+
+    Read all lines in a stream until EOF. If a user has passed a callback, call it for
+    each line.
+
+    Note that the lines are in bytes, since we can't assume that all command output will
+    be UTF-8 encoded. Higher level commands are advised to decode the output to Unicode,
+    if they know its encoding.
+    """
+    buf = b""
+    while True:
+        line = await sr.readline()
+        if sr.at_eof():
+            break
+        if callback is not None:
+            callback(line)
+        # TODO: This would be a good place to log the received line, mostly for debug
+        # logging.
+        buf += line
+    return buf
+
+
+async def run_command(
+    args: List[str],
+    *,
+    error_message: str,
+    timeout_message: str,
+    timeout: Optional[float],
+    stdout_callback: Optional[Callable] = None,
+    stderr_callback: Optional[Callable] = None,
+) -> Tuple[bytes, bytes]:
+    """Run a command and get its output.
+
+    Run a command using asyncio.subprocess, consume its standard streams, and return its
+    output in bytes.
+
+    :raises RuntimeError: if the process returns a non-zero exit status
+    :raises TimeoutError: if the process times out
+    """
+    # Start the provided command, and return a handle. The command will run in the
+    # background.
+    proc = await asyncio.subprocess.create_subprocess_exec(
+        *args,
+        stdout=asyncio.subprocess.PIPE,
+        stderr=asyncio.subprocess.PIPE,
+    )
+
+    assert proc.stdout is not None
+    assert proc.stderr is not None
+
+    # Create asynchronous tasks that will consume the standard streams of the command,
+    # and call callbacks if necessary.
+    stdout_task = asyncio.create_task(read_stream(proc.stdout, stdout_callback))
+    stderr_task = asyncio.create_task(read_stream(proc.stderr, stderr_callback))
+
+    # Wait until the command has finished, for a specific timeout. Then, verify that the
+    # command has completed successfully. In any other case, raise an exception.
+    try:
+        ret = await asyncio.wait_for(proc.wait(), timeout=timeout)
+    except asyncio.exceptions.TimeoutError:
+        raise TimeoutError(timeout_message)
+    if ret != 0:
+        raise RuntimeError(error_message)
+
+    # Wait until the tasks that consume the command's standard streams have exited as
+    # well, and return their output.
+    stdout = await stdout_task
+    stderr = await stderr_task
+    return (stdout, stderr)
+
+
+class DangerzoneConverter:
+    def __init__(self) -> None:
+        self.percentage: float = 0.0
+
+    def calculate_timeout(
```

Comment:
I think the idea is to try and do this client-side. But we can only do so when the containers code streams pages. But while that's not done, I think we can leave it as-is, no?

Reply:
Maybe this can be addressed in https://github.com/freedomofpress/dangerzone/issues/443?

Reply:
My thinking was that if we leave it as is, folks won't be able to disable timeouts on Qubes. Given that Qubes use disposable VMs for conversion, I'm afraid that timeouts may be more common in that platform.

We can address it in #443, if we make it part of 0.5.0. Have we committed to this though? I'm not sure.

That's why I've created a different issue instead (#446), with a narrower scope, so that we'll certainly tackle it in on 0.5.0. In the meantime, we can resolve this discussion and continue there.